types/react-timeago/index.d.ts

```diff
@@ -27,11 +27,11 @@ declare namespace ReactTimeago {
         nextFormatter?: Formatter
     ) => React.ReactNode;
 
-    interface ReactTimeagoProps<T extends React.ComponentType> {
+    interface ReactTimeagoProps<T extends React.ComponentType | keyof JSX.IntrinsicElements = 'time'> {
```

Comment:
@Philipp91 anecdotally, yes, using `keyof JSX.IntrinsicElements` is likely to be substantially slower than just using `string`.  Whether that affects the editing experience in practice is a judgement call.  You need to trade off the value of the extra type checking and IntelliSense from providing a list of acceptable strings against this extra cost.  As someone who doesn't use this library on a regular basis, I can't really make the call.

Reply:
I see. I (wrongly) thought the slowness was limited to that particular unit test case that the bot complained about.

As for the type checking / IntelliSense, I think that's applicable in two places: Firstly, when specifying the actual value of `T`, I think it doesn't matter much. Secondly, after `T` is chosen, checking that all the props passed to the components are actually valid props for `T` or `ReactTimeago` itself, that's quite important, as it can help discover subtle bugs.

If I replace `keyof JSX.IntrinsicElements` with just `string` as you proposed, then I think that wouldn't break the latter desirable kind of type checking, as once the user specifies `T` it can be checked. However, I instead get this error when compiling the library already:

```
ERROR: 45:62  expect  TypeScript@4.2 compile error:
Type 'T' does not satisfy the constraint '"symbol" | "object" | "big" | "link" | "small" | "sub" | "sup" | "track" | "progress" | "a" | "abbr" | "address" |
"area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | ... 156 more ... | (new (props: any) => Component<...>)'.
  Type 'string | ComponentClass<{}, any> | FunctionComponent<{}>' is not assignable to type '"symbol" | "object" | "big" | "link" | "small" | "sub" | "sup" | "track" | "progress" | "a" | "abbr" | "address" | "area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | ... 156 more ... | (new (props: any) => Component<...>)'.
    Type 'string' is not assignable to type '"symbol" | "object" | "big" | "link" | "small" | "sub" | "sup" | "track" | "progress" | "a" | "abbr" | "address" | "area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | ... 156 more ... | (new (props: any) => Component<...>)'.
      Type 'T' is not assignable to type '"mpath"'.
        Type 'string | ComponentClass<{}, any> | FunctionComponent<{}>' is not assignable to type '"mpath"'.
          Type 'string' is not assignable to type '"mpath"'.
```

I guess `React.ComponentProps<T>` isn't available when `T` can be an arbitrary `string`. Do you have any tips on how to work around this?

Presumably with a conditional type (sth like `T extends keyof JSX.IntrinsicElements ? ComponentProps<T> : never`) we'd be back at a slow (and more convoluted) solution. Dropping `ComponentProps` entirely leads to compile errors at the call site (because I *do* pass in some props for the underlying component). And replacing it with `any` seems ugly.

Reply:
Unfortunately, I'm not aware of a silver bullet that give you both perf and correctness in scenarios like this.  Personally, I tend to err on the side of using the more correct types until perf is shown to be a problem.  The difficulty, in this case, is that it will be hard for consumers to attribute any new editor delays they might see to this change, so there won't be a good way to react if the perf impact is excessiveness.  However, if you are or work with heavy users of this library, then your own experience should be a reasonably good indicator.

Reply:
> Personally, I tend to err on the side of using the more correct types until perf is shown to be a problem.

That does seem reasonable in general, and also applicable in this case. At least in my application, `TimeAgo` has only two call sites, unlike other components like `TextField` which have dozens. Even "heavy users" probably don't have more than a handful call sites. So the performance impact might not matter in practice.

So let's leave as is, for best type correctness.